### PR TITLE
Add curl-based functionality to download files from a URL.

### PR DESCRIPTION
**Pull request by Pentarctagon**
This is a complement to #7416 where it would be used to replace opening the replay's download URL in a browser.

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Hmm… seems fine, but I think I'd prefer to change all instances of "curl" to "http" (excepting actual Curl API stuff obviously). That means the namespace, the file, and the logdomain, maybe one or two other things.

**Comment by Pentarctagon** [comment]
It's not really a set of general http functionality though?

**Comment by CelticMinstrel** [comment]
It's not _generic_ HTTP functionality, but that is basically what it's doing, isn't it?

I'm not specifically attached to HTTP if you have something better; I would just prefer not encoding the library used into the names of functions that theoretically could later be rewritten to use something else.

Also, I have a feeling making it required is going to be really annoying for me, as the Mac compile stuff will need updating. Could it perhaps be made optional so I can at least continue to build while waiting for that?

**Comment by Pentarctagon** [comment]
I don't like how broad just "http" is, is the thing. Calling it "curl" may tie it to that specific library, but the name doesn't suggest a broader set of functionality than actually exists.

I don't plan to merge this until the MacCompileStuff is updated anyway, so I don't think it needs to be made optional.

**Comment by CelticMinstrel** [comment]
I'm not quite sure why HTTP implies such a broader set of functionality than what already exists (this is literally "using HTTP to download a file) and why "curl" would fail to imply the same broader set of functionality (handling HTTP is literally what curl is for, so shouldn't both have the same implication), but there are surely other options if you dislike it that much.

For example you could just call the logdomain "download" and drop any references to curl outside of the actual API usage.

**Comment by Pentarctagon** [comment]
Changed it to `network`.

File: src/curl_https_download_file.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <curl/curl.h>
+
+static lg::log_domain log_network("network");
+#define ERR_NW LOG_STREAM(err, log_network)
+#define DBG_NW LOG_STREAM(debug, log_network)
+
+namespace network
+{
+    static size_t write_callback(char* contents, size_t size, size_t nmemb, void* buffer)
```
Comment (Vultraz) [inline src/curl_https_download_file.cpp:28]: Does this need  to be `static`?

Comment (Pentarctagon) [reply src/curl_https_download_file.cpp:28]: Looks like it doesn't.

Comment (Pentarctagon) [reply src/curl_https_download_file.cpp]: Actually, apparently macOS requires it to be static.

**Comment by CelticMinstrel** [comment]
I said this in IRC too but I think the problem is in the Xcode file, not the environment. It looks like libcurl has not been added to a target definition.

**Comment by Pentarctagon** [comment]
I'll be merging this in a couple days barring any other comments, since the CI is passing now.

File: src/network_download_file.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+            CURLcode res = curl_easy_perform(curl);
+
+            if(res != CURLE_OK) {
+                ERR_NW << "Error downloading file from url `" << url << "`.\n"
+                    << "Short error: " << curl_easy_strerror(res) << "\n"
+                    << "Long error: " << std::string(error);
+            } else {
+                try {
+                    filesystem::write_file(local_path, buffer);
```
Comment (soliton-) [inline src/network_download_file.cpp:64]: Does this overwrite files without asking?

Comment (Pentarctagon) [reply src/network_download_file.cpp:64]: Yes.

File: src/network_download_file.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    void download(const std::string& url, const std::string& local_path)
+    {
+        CURL* curl = curl_easy_init();
+        std::string buffer;
+        // curl doesn't initialize the error buffer until version 7.60.0, which isn't currently available on all supported macOS versions
+        char error[CURL_ERROR_SIZE];
+        std::fill_n(error, CURL_ERROR_SIZE-1, ' ');
+        error[CURL_ERROR_SIZE-1] = '\0';
+
+        if(curl) {
```
[the remark this answers is not in the document]
Comment (Pentarctagon) [reply src/network_download_file.cpp:45]: Added in ccb7001

File: src/network_download_file.cpp (hunk 2 above)
Comment (Pentarctagon) [reply src/network_download_file.cpp:64]: Resolved in d280485

**Pull request merged**